File: android/app/src/main/kotlin/net/eduverseapp/platform/upload/UploadModels.kt
package net.eduverseapp.platform.upload

import org.json.JSONArray
import org.json.JSONObject

/**
 * A single upload job the native pipeline runs end-to-end:
 * init -> transfer (direct or multipart) -> complete -> callback.
 *
 * All fields needed to run WITHOUT the Flutter/Dart isolate are captured here
 * and persisted in [UploadStore] so the job survives app kill and device reboot.
 */
data class UploadJobData(
    val jobId: String,
    val filePath: String,
    val fileSize: Long,
    val title: String,
    // ── endpoints (already resolved on the Dart side) ──
    val initUrl: String,
    val completeUrl: String,
    val abortUrl: String,
    val callbackUrl: String,
    val callbackMethod: String,
    // JSON body sent to the init endpoint
    val initBody: String,
    // For the course endpoints that return data.thumbnail/data.video; else null
    val courseAssetKey: String?,
    // Callback body template with a __FILE_URL__ placeholder for the final url
    val callbackBodyTemplate: String,
    // For multipart complete/abort we always POST to completeUrl/abortUrl
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("jobId", jobId)
        put("filePath", filePath)
        put("fileSize", fileSize)
        put("title", title)
        put("initUrl", initUrl)
        put("completeUrl", completeUrl)
        put("abortUrl", abortUrl)
        put("callbackUrl", callbackUrl)
        put("callbackMethod", callbackMethod)
        put("initBody", initBody)
        put("courseAssetKey", courseAssetKey ?: JSONObject.NULL)
        put("callbackBodyTemplate", callbackBodyTemplate)
    }

    companion object {
        fun fromJson(o: JSONObject): UploadJobData = UploadJobData(
            jobId = o.getString("jobId"),
            filePath = o.getString("filePath"),
            fileSize = o.getLong("fileSize"),
            title = o.optString("title", "Upload"),
            initUrl = o.getString("initUrl"),
            completeUrl = o.getString("completeUrl"),
            abortUrl = o.getString("abortUrl"),
            callbackUrl = o.getString("callbackUrl"),
            callbackMethod = o.optString("callbackMethod", "POST"),
            initBody = o.getString("initBody"),
            courseAssetKey = if (o.isNull("courseAssetKey")) null else o.optString("courseAssetKey"),
            callbackBodyTemplate = o.getString("callbackBodyTemplate"),
        )
    }
}

/** Parsed init response (direct or multipart). */
data class InitResult(
    val isMultipart: Boolean,
    val uploadUrl: String?,
    val fileUrl: String,
    val key: String?,
    val s3UploadId: String?,
    val parts: List<PartUrl>,
)

data class PartUrl(val partNumber: Int, val uploadUrl: String)

/** Terminal result persisted for the Dart side to reconcile. */
data class UploadResult(
    val jobId: String,
    val status: String, // completed | failed
    val fileUrl: String?,
    val error: String?,
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("jobId", jobId)
        put("status", status)
        put("fileUrl", fileUrl ?: "")
        put("error", error ?: "")
        put("completedAt", System.currentTimeMillis())
    }
}

/** Parse the backend init envelope into an [InitResult]. */
object InitParser {
    fun parse(body: String, courseAssetKey: String?): InitResult? {
        return try {
            val root = JSONObject(body)
            // Unwrap { data: {...} }
            var d: JSONObject = root.optJSONObject("data") ?: root
            // Course endpoints nest one more level: data.data.{thumbnail|video}
            val nested = d.optJSONObject("data")
            if (nested != null && (nested.has("thumbnail") || nested.has("video"))) {
                val sectionKey = courseAssetKey ?: "thumbnail"
                nested.optJSONObject(sectionKey)?.let { d = it }
            }
            val isMultipart = d.optBoolean("isMultipart", false)
            val partsArr: JSONArray = d.optJSONArray("parts") ?: JSONArray()
            val parts = ArrayList<PartUrl>()
            for (i in 0 until partsArr.length()) {
                val p = partsArr.getJSONObject(i)
                parts.add(PartUrl(p.optInt("partNumber"), p.optString("uploadUrl")))
            }
            InitResult(
                isMultipart = isMultipart,
                uploadUrl = d.optString("uploadUrl").ifEmpty { null },
                fileUrl = d.optString("fileUrl", ""),
                key = d.optString("key").ifEmpty { null },
                s3UploadId = if (isMultipart) d.optString("uploadId").ifEmpty { null } else null,
                parts = parts,
            )
        } catch (e: Exception) {
            null
        }
    }

    fun extractFileUrl(body: String): String? {
        return try {
            val o = JSONObject(body)
            val data = o.optJSONObject("data")
            (data?.optString("fileUrl") ?: o.optString("fileUrl")).ifEmpty { null }
        } catch (_: Exception) {
            null
        }
    }
}

File: ios/Runner/UploadModels.swift
import Foundation

// MARK: - Upload Job Data (received from Dart)

struct UploadJobData: Codable {
    let jobId: String
    let filePath: String
    let fileSize: Int64
    let title: String
    let initUrl: String
    let completeUrl: String
    let abortUrl: String
    let callbackUrl: String
    let callbackMethod: String
    let initBody: String
    let courseAssetKey: String?
    let callbackBodyTemplate: String
    let createdAt: Int64

    init(from dict: [String: Any]) {
        jobId = dict["jobId"] as? String ?? ""
        filePath = dict["filePath"] as? String ?? ""
        fileSize = dict["fileSize"] as? Int64 ?? 0
        title = dict["title"] as? String ?? "Upload"
        initUrl = dict["initUrl"] as? String ?? ""
        completeUrl = dict["completeUrl"] as? String ?? ""
        abortUrl = dict["abortUrl"] as? String ?? ""
        callbackUrl = dict["callbackUrl"] as? String ?? ""
        callbackMethod = dict["callbackMethod"] as? String ?? "POST"
        initBody = dict["initBody"] as? String ?? "{}"
        courseAssetKey = dict["courseAssetKey"] as? String
        callbackBodyTemplate = dict["callbackBodyTemplate"] as? String ?? "{}"
        createdAt = dict["createdAt"] as? Int64 ?? Int64(Date().timeIntervalSince1970 * 1000)
    }

    func toDict() -> [String: Any] {
        var d: [String: Any] = [
            "jobId": jobId,
            "filePath": filePath,
            "fileSize": fileSize,
            "title": title,
            "initUrl": initUrl,
            "completeUrl": completeUrl,
            "abortUrl": abortUrl,
            "callbackUrl": callbackUrl,
            "callbackMethod": callbackMethod,
            "initBody": initBody,
            "callbackBodyTemplate": callbackBodyTemplate,
            "createdAt": createdAt,
        ]
        if let key = courseAssetKey { d["courseAssetKey"] = key }
        return d
    }
}

// MARK: - Init Response (parsed from backend)

struct InitResult {
    let isMultipart: Bool
    let uploadUrl: String?
    var fileUrl: String
    let key: String?
    let s3UploadId: String?
    var parts: [PartUrl]
}

struct PartUrl {
    let partNumber: Int
    var uploadUrl: String
}

// MARK: - Terminal Result (persisted for Dart to poll)

struct UploadResult: Codable {
    let jobId: String
    let status: String // "completed" or "failed"
    let fileUrl: String?
    let error: String?
    let completedAt: Int64

    init(jobId: String, status: String, fileUrl: String?, error: String?) {
        self.jobId = jobId
        self.status = status
        self.fileUrl = fileUrl
        self.error = error
        self.completedAt = Int64(Date().timeIntervalSince1970 * 1000)
    }

    func toDict() -> [String: Any] {
        var d: [String: Any] = [
            "jobId": jobId,
            "status": status,
            "completedAt": completedAt,
        ]
        if let url = fileUrl { d["fileUrl"] = url }
        if let err = error { d["error"] = err }
        return d
    }
}

// MARK: - Init Parser

struct InitParser {
    static func parse(_ body: String, courseAssetKey: String?) -> InitResult? {
        guard let data = body.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }

        var d: [String: Any] = (root["data"] as? [String: Any]) ?? root

        // Course endpoints nest: data.data.{thumbnail|video}
        if let nested = d["data"] as? [String: Any],
           (nested["thumbnail"] != nil || nested["video"] != nil) {
            let sectionKey = courseAssetKey ?? "thumbnail"
            if let section = nested[sectionKey] as? [String: Any] {
                d = section
            }
        }

        let isMultipart = d["isMultipart"] as? Bool ?? false
        let uploadUrl = d["uploadUrl"] as? String
        let fileUrl = d["fileUrl"] as? String ?? ""
        let key = d["key"] as? String
        let s3UploadId = isMultipart ? (d["uploadId"] as? String) : nil

        var parts: [PartUrl] = []
        if let partsArr = d["parts"] as? [[String: Any]] {
            for p in partsArr {
                if let num = p["partNumber"] as? Int {
                    parts.append(PartUrl(
                        partNumber: num,
                        uploadUrl: p["uploadUrl"] as? String ?? ""
                    ))
                }
            }
        }

        return InitResult(
            isMultipart: isMultipart,
            uploadUrl: uploadUrl,
            fileUrl: fileUrl,
            key: key,
            s3UploadId: s3UploadId,
            parts: parts
        )
    }

    static func extractFileUrl(_ body: String) -> String? {
        guard let data = body.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        if let dataObj = json["data"] as? [String: Any],
           let url = dataObj["fileUrl"] as? String, !url.isEmpty {
            return url
        }
        return json["fileUrl"] as? String
    }
}
